src/gathernodeinputhandler.cpp

```diff
@@ -0,0 +1,74 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "gathernodeinputhandler.hpp"
+
+#include "logging.hpp"
+#include "ov_utils.hpp"
+
+namespace ovms {
+
+GatherNodeInputHandler::GatherNodeInputHandler(uint32_t inputsMissingCount, session_id_t shardsCount) :
+    NodeInputHandler(inputsMissingCount) {
+    expectedDependencies *= shardsCount;
+}
+
+void GatherNodeInputHandler::setInput(const std::string& inputName, InferenceEngine::Blob::Ptr& ptr, session_id_t shardId) {
+    auto it = shardsStorage.find(inputName);
+    if (it == shardsStorage.end()) {
+        shard_map_t shardMap{{shardId, ptr}};
+        auto itDidInsertPair = shardsStorage.emplace(inputName, std::move(shardMap));
+        it = itDidInsertPair.first;
+        if (!itDidInsertPair.second) {
+            throw std::runtime_error("Tried to insert the same input twice with the same shard id");  // TODO recoverable error
+        }
+    } else {
+        it->second.emplace(shardId, ptr);
+    }
+}
+
+Status GatherNodeInputHandler::notifyFinishedDependency() {
+    NodeInputHandler::notifyFinishedDependency();
+    if (expectedDependencies > 0) {
+        return StatusCode::OK;
+    }
+    for (auto& [inputName, shardMap] : shardsStorage) {
+        const auto shardsCount = shardMap.size();
+        SPDLOG_LOGGER_DEBUG(dag_executor_logger, "Consolidating: {} shards for input: {}", shardsCount, inputName);
+        auto tensorDesc = shardMap[0]->getTensorDesc();
```

Comment:
Zero is misleading here - I had to check map type to see that it is in fact session key. Maybe declare the key before hand, to suggest this is session key?
```
session_key_t firstShardId = 0;
auto tensorDesc = shardMap[firstShardId]->getTensorDesc();
```
Are you sure 0 exists before you access it?

Reply:
As for key naming you are right this would be more readable. As for being sure if 0 exists:

At this point remainingDependencies should be equal 0 which means all shards were collected. If this wouldn't be correct statement here it would mean that GatherNodeInputHandler was constructed with wrong parameters.

NodeInputHandler takes predecessors count as an "remainingDependencies" & GatherNodeInputHandler additionally takes shardsCount and then multiply this "remainingDependencies".

I think we can rely here on other code parts working - which are unit tested.

When CustomNode will be fully functional we need to add test with multiple inputs coming from predecessor node into gather node. Then we will have coverage for potential future bugs introduced.

Reply:
Maybe throwing an exception if size is 0 would be good option

Reply:
I missed using operator[] here. I should be using .at().